### PR TITLE
 Conditional Concurrency Requirement Implementation

### LINKED ISSUE #44: Support conditional branch jobs based on the flag or output of previous job

### PR DESCRIPTION
**Pull request by SrinivasChilveri**
What this PR does / why we need it:
This PR adds new requirement Conditional Concurrency.
Conditional Concurrency Requirement Implementation
Conditional Concurrency Example
Conditional Concurrency E2E test case

Which issue(s) this PR fixes (optional, in fixes #(, fixes #<issue_number>, ...) format, will close the issue(s) when PR gets merged):

Fixes #44

### CODE REVIEW

File: cmd/genectl/parser/parser.go [hunk 1]
```diff
@@ -153,6 +156,17 @@ func InstantiateWorkflow(workflow *Workflow, inputs map[string]interface{}, tool
 			tmpJob.Commands = append(tmpJob.Commands, tool.Command)
 		}
 
+		if jobInfo.Condition != nil {
+			// populate data for condition
+			prefix := fmt.Sprintf("workflows.%s.condition", jobName)
+
+			cond, err := InstantiateCondition(prefix, jobInfo.Condition, inputsReplaceData)
```
Comment (hzxuzhonghu) [inline cmd/genectl/parser/parser.go]: `Condition` as a string satisfy if you pass the input data in to decide the condition output. If it is true/false, construct the execution accordingly.

And if it is `check_result(xxx, xxx)`, just pass down to execution.

Keep in mind, in api, we should make filed as simple as possible,  and also as concrete  as possible. Should not define a complicated api model in order to implement it easily.

Our target is to make users happy to use, and prevent causing confusion.

File: example/simple-sample-chkresult/README.md [hunk 2]
```diff
@@ -0,0 +1,21 @@
+## Overview
+
```
Comment (hzxuzhonghu) [inline example/simple-sample-chkresult/README.md]: why not rename `simple-sample-chkresult` to some thing like `conditionalxxxx`?

Comment (hzxuzhonghu) [reply example/simple-sample-chkresult/README.md]: I mean the dir name.

File: pkg/apis/gene/v1alpha1/types.go [hunk 3]
```diff
@@ -171,6 +176,10 @@ type Task struct {
 	// Specifies the dependency by this task
 	// +optional
 	Dependents []Dependent `json:"dependents"`
+
+	// Specifies the condition for this task
```
Comment (hzxuzhonghu) [inline pkg/apis/gene/v1alpha1/types.go]: more comments, `The task will be executed only when condition satisfied.`

File: pkg/apis/gene/v1alpha1/types.go [hunk 4]
```diff
@@ -114,6 +114,11 @@ type ExecutionSpec struct {
 	Parallelism *int64 `json:"parallelism,omitempty"`
 }
 
+// Condition  conditional branch handling  in Task
```
Comment (hzxuzhonghu) [inline pkg/apis/gene/v1alpha1/types.go]: // Condition conditional branch handling in Task

File: cmd/genectl/parser/condition_test.go [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+	"testing"
+)
+
+func TestIsCheckResultFunc(t *testing.T) {
+	testCases := []struct {
+		str    string
+		expect bool
+	}{
+		{
+			str:    "check_result(job-target, 'c')",
```
Comment (hzxuzhonghu) [inline cmd/genectl/parser/condition_test.go]: any difference between `check_result(job-target, 'c')` and `check_result(job-target, c)`?

File: cmd/genectl/parser/job.go [hunk 6]
```diff
@@ -230,3 +230,11 @@ func TransCommandIter2ExecCommandIter(commandsIter CommandsIter) *execv1alpha1.C
 	}
 	return &execCommandIter
 }
+
+func TransCond2ExecCond(condition interface{}) *execv1alpha1.Condition {
+	execCond := execv1alpha1.Condition{}
+
+	execCond.Condition = condition
```
Comment (hzxuzhonghu) [inline cmd/genectl/parser/job.go]: This should not be not like this.

File: pkg/controller/execution_job.go [hunk 7]
```diff
@@ -188,18 +188,49 @@ func (e *ExecutionJobController) syncHandler(event Event) error {
...
-					if err != nil {
-						return fmt.Errorf("getJobResult failed : %v", err)
+					// flag setting true to handle if Condition is nil
+					flag := true
+					var err error
+					if child.Data.DynamicJob.Condition != nil {
+						glog.V(2).Infof(" conditional based job condition:%v", child.Data.DynamicJob.Condition)
+						flag, err = e.evalConditionResult(vertex.Data.Job, child, graph, event.Key)
+						if err != nil {
+							return fmt.Errorf("evalConditionResult failed : %v", err)
```
Comment (hzxuzhonghu) [inline pkg/controller/execution_job.go]: why return? shouldn't be continue? IIUC, You should continue for other children?

File: pkg/controller/execution_job.go [hunk 8]
```diff
@@ -188,18 +188,49 @@ func (e *ExecutionJobController) syncHandler(event Event) error {
...
+						}
+						if !flag {
+							// if the condition or check_result validation is false
+							// then we don't create the k8s job and make the job is Finished true
+							// so that other jobs will continue or execution will complete
+							glog.V(2).Infof(" The final condition is false")
+							child.Data.Finished = true
+							return nil
```
Comment (hzxuzhonghu) [inline pkg/controller/execution_job.go]: same here?

File: pkg/controller/execution_job.go [hunk 9]
```diff
@@ -188,18 +188,49 @@ func (e *ExecutionJobController) syncHandler(event Event) error {
...
+						}
 					}
 
-					// construct the dynamic job based on result
-					err = e.createDynamicJob(child, result, graph, event.Key)
-					if err != nil {
-						return fmt.Errorf("createDynamicJob failed : %v", err)
+					if flag && len(child.Data.DynamicJob.CommandSet) > 0 && child.Data.DynamicJob.CommandsIter == nil {
```
Comment (hzxuzhonghu) [inline pkg/controller/execution_job.go]: no need to check flag , above logic guaranteed it be true

Comment (SrinivasChilveri) [reply pkg/controller/execution_job.go]: agreed

File: pkg/controller/execution_job.go (hunk 7 above)
Comment (SrinivasChilveri) [reply pkg/controller/execution_job.go]: yeah, we can continue

File: pkg/controller/execution_job.go (hunk 8 above)
Comment (SrinivasChilveri) [reply pkg/controller/execution_job.go]: yeah, we can continue

File: example/simple-sample-chkresult/README.md (hunk 2 above)
Comment (SrinivasChilveri) [reply example/simple-sample-chkresult/README.md]: chnaged

File: pkg/apis/gene/v1alpha1/types.go (hunk 3 above)
Comment (SrinivasChilveri) [reply pkg/apis/gene/v1alpha1/types.go]: added

File: cmd/genectl/parser/condition_test.go (hunk 5 above)
Comment (SrinivasChilveri) [reply cmd/genectl/parser/condition_test.go]: it allows only " " or '  ' doesn't allow check_result(job-target, c) as it fails at regexp matching it self. I hope with the enhancement which we planned we will over come all these issue.

File: cmd/genectl/parser/job.go (hunk 6 above)
Comment (SrinivasChilveri) [reply cmd/genectl/parser/job.go]: Its interface but the data representation is different in client side & server side. any how we are planning to update this feature with other possibilities like exist ,In,NotIn etc as k8s.                    
 what do you say?

**Comment by hzxuzhonghu** [comment]
ok， let's make it in first.

**Pull request merged**